### ISSUE TITLE
[Android] java.lang.UnsatisfiedLinkError in Android-x86-64

### ISSUE THREAD

**Issue by phj233** [aldenml/libtorrent4j#284]
I use libtorrent4j in my Android project, and when I debug it in Android Studio's x86_64 AVD, the following error appears:

```kotlin
2024-12-08 13:06:15.972 11824-11862 nativeloader            top.phj233.magplay.debug             D  Load /data/app/~~786TCTz4k5O20OVUujDzkw==/top.phj233.magplay.debug-UCSMUp7mxK-HZpFG-BSuNw==/lib/x86_64/libtorrent4j.so using ns clns-7 from class loader (caller=/data/app/~~786TCTz4k5O20OVUujDzkw==/top.phj233.magplay.debug-UCSMUp7mxK-HZpFG-BSuNw==/base.apk!classes4.dex): dlopen failed: empty/missing DT_HASH/DT_GNU_HASH in "/data/app/~~786TCTz4k5O20OVUujDzkw==/top.phj233.magplay.debug-UCSMUp7mxK-HZpFG-BSuNw==/lib/x86_64/libtorrent4j.so" (new hash type from the future?)
2024-12-08 13:06:16.105 11824-11824 AndroidRuntime          top.phj233.magplay.debug             E  FATAL EXCEPTION: main (Ask Gemini)
                                                                                                    Process: top.phj233.magplay.debug, PID: 11824
                                                                                                    java.lang.LinkageError: Look for your architecture binary instructions at: https://github.com/aldenml/libtorrent4j
                                                                                                    	at org.libtorrent4j.swig.libtorrent_jni.<clinit>(libtorrent_jni.java:28)
                                                                                                    	at org.libtorrent4j.swig.alert.<clinit>(alert.java:546)
                                                                                                    	at org.libtorrent4j.alerts.AlertType.<clinit>(AlertType.java:18)
                                                                                                    	at org.libtorrent4j.SessionManager.<clinit>(SessionManager.java:54)
                                                                                                    	at top.phj233.magplay.torrent.TorrentManager$magnetLinkParser$2.invokeSuspend(TorrentManager.kt:29)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:101)
                                                                                                    	at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:113)
                                                                                                    	at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:89)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:589)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:823)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:720)
                                                                                                    	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:707)
                                                                                                    	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@410542c, Dispatchers.Main.immediate]
                                                                                                    Caused by: java.lang.UnsatisfiedLinkError: dlopen failed: empty/missing DT_HASH/DT_GNU_HASH in "/data/app/~~786TCTz4k5O20OVUujDzkw==/top.phj233.magplay.debug-UCSMUp7mxK-HZpFG-BSuNw==/lib/x86_64/libtorrent4j.so" (new hash type from the future?)
                                                                                                    	at java.lang.Runtime.loadLibrary0(Runtime.java:1081)
                                                                                                    	at java.lang.Runtime.loadLibrary0(Runtime.java:1003)
                                                                                                    	at java.lang.System.loadLibrary(System.java:1765)
                                                                                                    	at org.libtorrent4j.swig.libtorrent_jni.<clinit>(libtorrent_jni.java:23)
                                                                                                    	... 12 more
```

```
// app: build.gradle.kts
plugins {
    alias(libs.plugins.android.application)
    alias(libs.plugins.kotlin.android)
    alias(libs.plugins.kotlin.serialization)
    alias(libs.plugins.kotlin.ksp)
    alias(libs.plugins.compose.compiler)
}

android {
    namespace = "top.phj233.magplay"
    compileSdk = 35


    defaultConfig {
        applicationId = "top.phj233.magplay"
        targetSdk = 35
        minSdk = 26
        versionCode = 1
        versionName = "1.0"

        multiDexEnabled = true
        testInstrumentationRunner = "androidx.test.runner.AndroidJUnitRunner"
        ksp {
            arg("room.schemaLocation", "$projectDir/schemas")
        }
    }


    splits {
        abi {
            isEnable = true
            reset()
            include("arm64-v8a", "armeabi-v7a", "x86", "x86_64")
            isUniversalApk = true
        }
    }

    buildTypes {
        debug {
            isMinifyEnabled = false
            isShrinkResources = false
            proguardFiles(
                getDefaultProguardFile("proguard-android-optimize.txt"),
                "proguard-rules.pro"
            )
            applicationIdSuffix = ".debug"
        }
        release {
            isMinifyEnabled = true
            proguardFiles(
                getDefaultProguardFile("proguard-android-optimize.txt"),
                "proguard-rules.pro"
            )
        }
    }
    compileOptions {
        sourceCompatibility = JavaVersion.VERSION_17
        targetCompatibility = JavaVersion.VERSION_17
    }

    kotlinOptions {
        jvmTarget = "17"
    }
    buildFeatures {
        compose = true
    }
    composeOptions {
        kotlinCompilerExtensionVersion = "1.5.1"
    }
    packaging {
        resources {
            excludes += "/META-INF/{AL2.0,LGPL2.1}"
        }

        jniLibs {
            useLegacyPackaging = true
        }
        dex {
            useLegacyPackaging = true
        }
    }

}

dependencies {
    implementation(libs.androidx.core.ktx)
    implementation(libs.androidx.lifecycle.runtime.ktx)
    implementation(libs.androidx.activity.compose)
    implementation(libs.androidx.navigation.compose)
    implementation(platform(libs.androidx.compose.bom))
    implementation(libs.androidx.ui)
    implementation(libs.androidx.ui.graphics)
    implementation(libs.androidx.ui.tooling.preview)
    implementation(libs.androidx.material3)
    implementation(libs.androidx.material.icons.extended)
    // tencent-mmkv
    implementation(libs.tencent.mmkv)
    // kotlin-serialization-json
    implementation(libs.kotlin.serilization.json)
    // room
    implementation(libs.androidx.room)
    implementation(libs.androidx.room.ktx)
    implementation(libs.androidx.documentfile)
    annotationProcessor(libs.androidx.room.compiler)
    ksp(libs.androidx.room.compiler)
    //media3
    implementation(libs.media3.exoplayer)
    implementation(libs.media3.common)
    implementation(libs.media3.ui)
    implementation(libs.media3.session)
    // koin
    implementation(libs.koin.android)
    implementation(libs.koin.compose)
    implementation(libs.koin.compose.viewmodel)
    implementation(libs.koin.core)
    // libtorrent4j
    implementation(libs.libtorrent4j.android.arm)
    implementation(libs.libtorrent4j.android.arm64)
    implementation(libs.libtorrent4j.android.x86)
    implementation(libs.libtorrent4j.android.x64)
    // ktor
    implementation(libs.ktor.client.core)
    implementation(libs.ktor.client.cio)
    // test
    testImplementation(libs.junit)
    androidTestImplementation(libs.androidx.junit)
    androidTestImplementation(libs.androidx.espresso.core)
    androidTestImplementation(platform(libs.androidx.compose.bom))
    androidTestImplementation(libs.androidx.ui.test.junit4)
    debugImplementation(libs.androidx.ui.tooling)
    debugImplementation(libs.androidx.ui.test.manifest)
}
```

When I use my phone(arm x64) to debug, the above errors will not appear . And I have introduced the following dependencies, proguard-pro also added related content

```toml
# libs.versions.toml:
# libtorrent4j = "2.1.0-35"
# libtorrent4j 
libtorrent4j-android-arm = { group = "org.libtorrent4j", name = "libtorrent4j-android-arm", version.ref = "libtorrent4j" }
libtorrent4j-android-arm64 = { group = "org.libtorrent4j", name = "libtorrent4j-android-arm64", version.ref = "libtorrent4j" }
libtorrent4j-android-x86 = { group = "org.libtorrent4j", name = "libtorrent4j-android-x86", version.ref = "libtorrent4j" }
libtorrent4j-android-x64 = { group = "org.libtorrent4j", name = "libtorrent4j-android-x86_64", version.ref = "libtorrent4j" }
```

```properties
# libtorrent4j proguard-pro
-keep class org.libtorrent4j.** { *; }
```

And my code cannot parse the magnet. I don't think there is any problem with the coding. Can you help me solve this problem? The following method is to obtain metadata about magnet：

```kotlin
object TorrentManager {

    suspend fun magnetLinkParser(magnetLink: String): Result<MagPlayTorrentInfo> = withContext(Dispatchers.IO) {
        try {
            val deferred = CompletableDeferred<MagPlayTorrentInfo>()
            val session = SessionManager(true)

            val sp = SessionParams(SettingsPack().apply {
                setString(settings_pack.string_types.user_agent.swigValue(), "MagPlay/0.0.1")
                setBoolean(settings_pack.bool_types.announce_to_all_trackers.swigValue(), true)
                setBoolean(settings_pack.bool_types.dht_restrict_search_ips.swigValue(), false)
                setBoolean(settings_pack.bool_types.dht_restrict_routing_ips.swigValue(), false)
            })

            Log.d("TorrentParser", "magnetLink: $magnetLink")
            val updatedMagnet = mergeTracker(magnetLink, getTrackerList())
            Log.d("TorrentParser", "updatedMagnet: $updatedMagnet")

            session.addListener(object : AlertListener {
                override fun types(): IntArray? = null

                override fun alert(alert: Alert<*>?) {
                    when (alert?.type()) {
                        AlertType.METADATA_RECEIVED -> {
                            val th = (alert as MetadataReceivedAlert).handle()
                            val info = th.torrentFile()
                            val result = MagPlayTorrentInfo(
                                info.name(),
                                info.totalSize(),
                                (0 until info.numFiles()).map {
                                    TorrentFile(
                                        info.files().fileName(it),
                                        info.files().filePath(it),
                                        info.files().fileSize(it),
                                    )
                                },
                                info.infoHash().toString()
                            )
                            session.stop()
                            deferred.complete(result)
                        }
                        else -> {

                        }
                    }
                }
            })

            session.start(sp)
            session.startDht()

            // 设置超时
            withTimeoutOrNull(30000) {
                Result.success(deferred.await())
            } ?: run {
                session.stop()
                Result.failure(Exception("解析超时"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    private suspend fun getTrackerList():List<String> {
        val trackerList = mutableListOf<String>()
        val file = File(Environment.getExternalStoragePublicDirectory(
            Environment.DIRECTORY_DOWNLOADS
        ), "trackers.txt")
        if (!file.exists()) {
            // 使用ktor 请求下载https://cdn.jsdmirror.com/gh/XIU2/TrackersListCollection/best.txt
            // 并保存到 trackers.txt
            val client = HttpClient()
            val response = client.get("https://cdn.jsdmirror.com/gh/XIU2/TrackersListCollection/best.txt")
            response.bodyAsText().let {
                file.writeText(it)
            }
        }
        file.forEachLine {
            if (it.isBlank()) {
                return@forEachLine
            }
            trackerList.add(URLEncoder.encode(it, "UTF-8"))
        }
        return trackerList
    }

    // 合并tracker 和 magnetLink
    fun mergeTracker(magnetLink: String, trackerList: List<String>): String {
        val tracker = trackerList.joinToString("&tr=") { it }
        return "$magnetLink&tr=$tracker"
    }
 }
```

**Comment by aldenml** [comment]
this seems related to the recent changes about the 16 KB ELF alignment. Can you try another AVD, x86_64, but less advanced?

**Comment by phj233** [comment]
> this seems related to the recent changes about the 16 KB ELF alignment. Can you try another AVD, x86_64, but less advanced?

Thank you for your reply! After I tried using AVD with APIs 33 and 34, the problem with java.lang.UnsatisfiedLinkError disappeared. 
And I found the cause of the second problem. I forgot to use the ` fetchMagnet () ` method. If I want to parse a magnet, do I have to use ` fetchMagnet () ` after ` session.start() `? Is there any other way to just get metadata?
